Makefile

```diff
@@ -1682,3 +1682,10 @@ rustup-install-target-toolchain: rustup-set-version
 .PHONY: changelog
 changelog:
 	@BASE_BRANCH=$(BASE_BRANCH) BASE_TAG=$(BASE_TAG) ./build.assets/changelog.sh
+
+# check-gitleaks will run a gitleaks container to detect any secrets staged for commit
+# to prevent accidental leaks.
+.PHONE: check-gitleaks
+check-gitleaks: GITLEAKS_VERSION = v8.18.4
+check-gitleaks:
+	@docker run -v $(PWD):/path ghcr.io/gitleaks/gitleaks:$(GITLEAKS_VERSION) protect --verbose --redact --staged --source="/path"
```

Comment:
Yes - IMO that's a reasonable expectation. Most of our local tooling currently relies on Docker anyway.

Reply:
We could - but we want to pin the version, and running as a binary installed on every dev machine makes this very painful to cover in every case (multiple arches, multiple OSs, tool version upgrades and downgrades based off of Teleport major version branch).

Reply:
Given that this impacts all devs, it might be worth polling people in Slack to see how impactful this would be. What do you all think?

Unfortunately I don't think we have well-defined list of what devs are expected to be running and have installed. We used to have one maintained by IT as a part of the new hire checklist/setup, but I couldn't find it when I went looking for it last month.

Reply:
It's important to also note that pre-commit hooks are an opt-in feature of git and are configured on the client side. Everything here is more about setting expectations for developers and giving them a useful configuration for preventing accidental configuration leaking.

I could update the script to prefer a local `gitleaks` if present but to default to using the docker container with the pinned version.

Reply:
@doggydogworld Making this opt-in limits the usability quite a bit IMO - majority will never configure it out of their own volition and we'll have the same problem. Is there a way to make this mandatory? E.g. can we commit files that "lefthook install" command creates to the repository or something, would that work?

Reply:
@r0mant Unfortunately by design **pre-commit** hooks are opt-in on the clients side. Git does have **server-side** hooks but because GitHub is a cloud service these aren't supported because it's arbitrary code running on their servers. I could be wrong on the GitHub side of things nowadays since I haven't had to admin an org in a while and the documentation can be tricky because it is supported in GitHub Enterprise.

A post-commit job that resets the history and delete the offending commit is a good way to enforce it server side for GitHub. It has a bad dev UX though because their local history will no longer match what's on the server. The **pre-commit** hooks were designed so that devs are notified as early as possible if there is a problem before having it get rejected down the line when they push

Reply:
I don't think hosted Github Enterprise supports pre-receive hooks. We actually use Github's builtin [secret scanning](https://docs.github.com/en/enterprise-cloud@latest/code-security/secret-scanning/push-protection-for-repositories-and-organizations) with push protection enabled. @doggydogworld Do you know if there's a big overlap between what Github and gitleaks detect? Github has a list of supported partners [here](https://docs.github.com/en/enterprise-cloud@latest/code-security/secret-scanning/secret-scanning-patterns#supported-secrets).

Reply:
There is an overlap in some of the features but GitHub's secret scanning only works when you push. A push will contain history and if that history contains a secret then you will have to manually overwrite that history to clean it.

Pre-commit is intended to provide a better developer experience by notifying them before their commits get into a bad state. In this case gitleaks is meant to be complementary to GitHub's secret scanning.